great_expectations/datasource/fluent/config_str.py

```diff
@@ -72,6 +88,125 @@ def __get_validators__(cls):
         yield cls.validate
 
 
+UriParts: TypeAlias = Literal[  # https://docs.pydantic.dev/1.10/usage/types/#url-properties
+    "scheme", "host", "user", "password", "port", "path", "query", "fragment", "tld"
+]
+
+
+class UriPartsDict(TypedDict, total=False):
+    scheme: str
+    user: str | None
+    password: str | None
+    ipv4: str | None
+    ipv6: str | None
+    domain: str | None
+    port: str | None
+    path: str | None
+    query: str | None
+    fragment: str | None
+
+
+class ConfigUri(AnyUrl, ConfigStr):  # type: ignore[misc] # Mixin "validate" signature mismatch
+    """
+    Special type that enables great_expectation config variable substitution for the
+    `user` and `password` section of a URI.
+
+    Example:
+    ```
+    "snowflake://${MY_USER}:${MY_PASSWORD}@account/database/schema/table"
+    ```
+
+    Note: this type is meant to used as part of pydantic model.
+    To use this outside of a model see the pydantic docs below.
+    https://docs.pydantic.dev/usage/models/#parsing-data-into-a-specified-type
+    """
+
+    ALLOWED_SUBSTITUTIONS: ClassVar[set[UriParts]] = {"user", "password"}
+
+    min_length: int = 1
+    max_length: int = 2**16
+
+    def __init__(  # noqa: PLR0913 # for compatibility with AnyUrl
+        self,
+        template_str: str,
+        *,
+        scheme: str,
+        user: Optional[str] = None,
+        password: Optional[str] = None,
+        host: Optional[str] = None,
+        tld: Optional[str] = None,
+        host_type: str = "domain",
+        port: Optional[str] = None,
+        path: Optional[str] = None,
+        query: Optional[str] = None,
+        fragment: Optional[str] = None,
+    ) -> None:
+        if template_str:  # may have already been set in __new__
+            self.template_str: str = template_str
```

Comment:
A bit confused by this. Under what circumstances would `self.template_str` not be set by the time we get here?